executor/mppcoordmanager/mpp_coordinator_manager.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mppcoordmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/mpp"
+	"github.com/pingcap/tidb/executor/metrics"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/copr"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// InstanceMPPCoordinatorManager is a local instance mpp coordinator manager
+var InstanceMPPCoordinatorManager = newMPPCoordinatorManger()
+
+const (
+	detectFrequency = 5 * time.Minute
+)
+
+// CoordinatorUniqueID identifies a unique coordinator
+type CoordinatorUniqueID struct {
+	MPPQueryID kv.MPPQueryID
+	GatherID   uint64
+}
+
+// MPPCoordinatorManager manages all mpp coordinator instances
+type MPPCoordinatorManager struct {
+	mu             sync.Mutex
+	serverOn       bool
+	serverAddr     string // empty if server is off
+	coordinatorMap map[CoordinatorUniqueID]kv.MppCoordinator
+	wg             sync.WaitGroup
+	ctx            context.Context
+	cancel         context.CancelFunc
+	maxLifeTime    uint64 // in Nano
+}
+
+// Run use a loop to detect and remove out of time Coordinators
+func (m *MPPCoordinatorManager) Run() {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	m.wg.Add(1)
+	m.maxLifeTime = uint64(copr.TiFlashReadTimeoutUltraLong.Nanoseconds() + detectFrequency.Nanoseconds())
+	go func() {
+		defer m.wg.Done()
+		ticker := time.NewTicker(detectFrequency)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-m.ctx.Done():
+				return
+			case <-ticker.C:
+				m.detectAndDelete(uint64(time.Now().UnixNano()))
+			}
+		}
+	}()
+}
+
+func (m *MPPCoordinatorManager) detectAndDelete(nowTs uint64) {
+	var outOfTimeIDs []CoordinatorUniqueID
+	m.mu.Lock()
+	for id := range m.coordinatorMap {
+		if nowTs > id.MPPQueryID.QueryTs+m.maxLifeTime {
```

Comment:
Is it possible to delete only the coodinator that has received all the data but has not received the report summary?
I am a little worried that the long-distance mpp query will be deleted by mistake, such as the sql running a large etl

Reply:
Or does it follow the time of sql execution timeout?

Reply:
Yeah, that's the longest time one mpp task can take.

Reply:
It's hard coded here: https://github.com/pingcap/tidb/blob/9fc80a922a340d851d69327dc3de189bdb984fb3/store/copr/batch_coprocessor.go#L1133, and passed as timeout parameter to client-go sendRequest

Reply:
ok

Reply:
But this is not the max running time of a mpp query, this timeout means the client does not receive any data for `readTimeoutUltraLong` time, if a mpp query keep return data for every 1 minutes, it never timeout

Reply:
If that's the case, then I suggest only deleting the coordinator that received the data but did not receive all report status messages.

Reply:
I see, I'll use the coordinator's close flag as a double check.

Reply:
Done